source/common/upstream/cluster_manager_impl.cc

```diff
@@ -154,6 +154,38 @@ void ClusterManagerInitHelper::setInitializedCb(std::function<void()> callback)
   }
 }
 
+void ClusterManagerImpl::initializeClustersFromV1Json(const Json::Object& config) {
+  envoy::api::v2::ConfigSource cds_config;
+  if (config.hasObject("cds")) {
+    envoy::api::v2::Cluster cds_cluster;
+    Config::CdsJson::translateCluster(*config.getObject("cds")->getObject("cluster"), sds_config_,
+                                      cds_cluster);
+    loadCluster(cds_cluster, false);
+    Config::Utility::translateCdsConfig(*config.getObject("cds"), cds_config);
+    // We can now potentially create the CDS API once the backing cluster exists.
+    cds_api_ = factory_.createCds(cds_config, sds_config_, *this);
+    init_helper_.setCds(cds_api_.get());
+  } else {
+    init_helper_.setCds(nullptr);
+  }
+
+  for (const Json::ObjectSharedPtr& cluster_config : config.getObjectArray("clusters")) {
+    envoy::api::v2::Cluster cluster;
+    Config::CdsJson::translateCluster(*cluster_config, sds_config_, cluster);
+    loadCluster(cluster, false);
+  }
+}
+
+void ClusterManagerImpl::initializeClustersFromV2Proto(const envoy::api::v2::Bootstrap& bootstrap) {
+  for (const auto& cluster : bootstrap.bootstrap_clusters()) {
+    loadCluster(cluster, false);
+  }
+
+  // We can now potentially create the CDS API once the backing cluster exists.
```

Comment:
It's mandatory in the sense that you need to specify "CDS", but this can just be a file which we subscribe to. The main difference from v1 is that the clusters will be in a distinct file to the bootstrap file.

Do we have a need to consolidate these into a single file? From a implementation simplicity perspective, separate files were easier, as each file maps 1:1 to a subscription source.